addons/hr_expense/models/hr_expense.py

```diff
@@ -197,6 +201,59 @@ def _compute_employee_id(self):
             for expense in self:
                 expense.employee_id = self.env.user.with_company(expense.company_id).employee_id
 
+    @api.depends('employee_id', 'product_id', 'total_amount')
+    def _compute_duplicate_expense_ids(self):
+        self.duplicate_expense_ids = False
+
+        expenses = self.filtered(lambda e: e.employee_id and e.product_id and e.total_amount)
+        if expenses.ids:
+            duplicates_query = """
+                WITH expenses AS (
+                    SELECT employee_id, product_id, date, total_amount, company_id, currency_id
+                      FROM hr_expense
+                     WHERE id IN %(expense_ids)s
+                     GROUP BY employee_id, product_id, date, total_amount, company_id, currency_id)
+              SELECT ARRAY_AGG(he.id)
+                FROM hr_expense AS he
+               INNER JOIN expenses AS ex
+                  ON he.employee_id = ex.employee_id
+                 AND he.product_id = ex.product_id
+                 AND he.date = ex.date
+                 AND he.total_amount = ex.total_amount
+                 AND he.company_id = ex.company_id
+                 AND he.currency_id = ex.currency_id
+               GROUP BY he.employee_id, he.product_id, he.date, he.total_amount, he.company_id, he.currency_id
+              HAVING COUNT(id) > 1
+            """
+            self.env.cr.execute(duplicates_query, {
+                'expense_ids': tuple(expenses.ids),
+            })
+            duplicates = [x[0] for x in self.env.cr.fetchall()]
+
+            for ids in duplicates:
+                exp = expenses.filtered(lambda e: e.id in ids)
+                exp.duplicate_expense_ids = [(6, 0, ids)]
+                expenses = expenses - exp
+
+        new_expenses = expenses.filtered(lambda e: not e.id)
```

Comment:
@arj-odoo instead of duplicating logic, you could have `flush()` and have new records handled with your SQL query.

Let's say records A, B and C are duplicated. A and B are in the database, while C is new. A will see B as duplicate but not C, while C will see A and B as duplicates.